R/incidence.R

```diff
@@ -37,109 +36,85 @@ graph.incidence <- function(incidence, directed = FALSE, mode = c("all", "out",
 ##
 ## -----------------------------------------------------------------
 
-graph.incidence.sparse <- function(incidence, directed, mode, multiple,
-                                   weighted) {
-  n1 <- nrow(incidence)
-  n2 <- ncol(incidence)
+# Helper function to process sparse matrices (matrix to edgelist)
+process.sparse <- function(incidence, num_rows) {
   el <- mysummary(incidence)
-  el[, 2] <- el[, 2] + n1
+  el[, 2] <- el[, 2] + num_rows
+  as.matrix(el)
+}
 
-  if (!is.null(weighted)) {
+# Helper function to process dense matrices (matrix to edgelist)
+process.dense <- function(incidence, num_rows) {
+  nz_ids <- which(incidence != 0, arr.ind = TRUE)
+  nz_ids <- nz_ids[order(nz_ids[, 1], nz_ids[, 2]), , drop = FALSE]
+  el <- cbind(nz_ids, incidence[nz_ids])
+  el[, 2] <- el[, 2] + num_rows
+  el
+}
 
-    if (!directed || mode == 1) {
-      ## nothing do to
-    } else if (mode == 2) {
-      el[, 1:2] <- el[, c(2, 1)]
-    } else if (mode == 3) {
-      reversed_el <- el[, c(2, 1, 3)]
-      names(reversed_el) <- names(el)
-      el <- rbind(el, reversed_el)
-    }
+# adjust edgelist according to directionality of edges
+adjust.directionality <- function(el, mode, directed) {
+  if (!directed || mode == "out") {
+    # No adjustment needed
+    return(el)
+  } else if (mode == "in") {
```

Comment:
Can this be simplified further?

```suggestion
  }
  reversed_edges <- ...
  if (mode == "in") {
```

Reply:
Yes, I think this is probably the simplest solution?
```r
adjust.directionality <- function(el, mode, directed) {
  if (!directed || mode == "out") {
    # No adjustment needed
    return(el)
  }
  reversed_edges <- el[, c(2, 1, 3)]
  if (mode == "in") {
    return(reversed_edges)
  }
  if (mode %in% c("all", "total")) {
    # Add reversed edges
    return(rbind(el, reversed_edges))
  }
}
```

Reply:
Yeah, the second and third `return()` aren't even needed.

https://style.tidyverse.org/functions.html#return

Reply:
First and second are needed, otherwise nothing is returned for `mode == "in"`
``` r
el <- matrix(
  c(1, 3, 1, 3, 1, 2, 3, 4, 4, 5, 5, 8, 8, 8, 1, 2, 1, 2, 2, 2, 1),
  nrow = 7L,
  ncol = 3L,
  dimnames = list(NULL, c("i", "j", "x"))
)

el
#>      i j x
#> [1,] 1 4 1
#> [2,] 3 4 2
#> [3,] 1 5 1
#> [4,] 3 5 2
#> [5,] 1 8 2
#> [6,] 2 8 2
#> [7,] 3 8 1

adjust.directionality <- function(el, mode, directed) {
  if (!directed || mode == "out") {
    # No adjustment needed
    return(el)
  }
  reversed_edges <- el[, c(2, 1, 3)]
  if (mode == "in") {
    reversed_edges
  }
  if (mode %in% c("all", "total")) {
    # Add reversed edges
    rbind(el, reversed_edges)
  }
}

adjust.directionality(el,mode = "in",directed = TRUE)
adjust.directionality(el,mode = "out",directed = TRUE)
#>      i j x
#> [1,] 1 4 1
#> [2,] 3 4 2
#> [3,] 1 5 1
#> [4,] 3 5 2
#> [5,] 1 8 2
#> [6,] 2 8 2
#> [7,] 3 8 1
adjust.directionality(el,mode = "all",directed = TRUE)
#>       i j x
#>  [1,] 1 4 1
#>  [2,] 3 4 2
#>  [3,] 1 5 1
#>  [4,] 3 5 2
#>  [5,] 1 8 2
#>  [6,] 2 8 2
#>  [7,] 3 8 1
#>  [8,] 4 1 1
#>  [9,] 4 3 2
#> [10,] 5 1 1
#> [11,] 5 3 2
#> [12,] 8 1 2
#> [13,] 8 2 2
#> [14,] 8 3 1
```

<sup>Created on 2025-01-16 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>